flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/RetryExtension.java

```diff
@@ -0,0 +1,131 @@
+            ExtensionContext context) {
+        Method testMethod = context.getRequiredTestMethod();
+        RetryOnFailure retryOnFailure = testMethod.getAnnotation(RetryOnFailure.class);
+        RetryOnException retryOnException = testMethod.getAnnotation(RetryOnException.class);
+
+        if (retryOnFailure == null && retryOnException == null) {
+            // if nothing is specified on the test method, fall back to annotations on the class
+            retryOnFailure = context.getTestClass().get().getAnnotation(RetryOnFailure.class);
+            retryOnException = context.getTestClass().get().getAnnotation(RetryOnException.class);
+        }
+        // sanity check that we don't use both annotations
+        if (retryOnFailure != null && retryOnException != null) {
+            throw new IllegalArgumentException(
+                    "You cannot combine the RetryOnFailure and RetryOnException annotations.");
+        }
+
+        Map<String, Boolean> testLog =
+                (Map<String, Boolean>)
+                        context.getStore(RETRY_NAMESPACE)
+                                .getOrComputeIfAbsent(RETRY_KEY, key -> new HashMap<>());
+        testLog.put(getTestMethodKey(context), true);
+        if (retryOnFailure == null) {
+            int totalTimes = retryOnException.times() + 1;
+            RetryOnException finalRetryOnException = retryOnException;
+            return IntStream.rangeClosed(1, totalTimes)
+                    .mapToObj(
+                            i ->
+                                    new RetryContext(
+                                            i, totalTimes, finalRetryOnException.exception()));
+        } else {
+            int totalTimes = retryOnFailure.times() + 1;
+            return IntStream.rangeClosed(1, totalTimes)
+                    .mapToObj(i -> new RetryContext(i, totalTimes, null));
+        }
+    }
+
+    @Override
+    public void afterAll(ExtensionContext context) throws Exception {
+        context.getStore(RETRY_NAMESPACE).remove(RETRY_KEY);
+    }
+
+    public static String getTestMethodKey(ExtensionContext context) {
+        return context.getRequiredTestClass().getCanonicalName()
+                + "."
+                + context.getRequiredTestMethod().getName();
+    }
+
+    class RetryContext implements TestTemplateInvocationContext {
+        int retryIndex;
+        int totalTimes;
+        Class<? extends Throwable> repeatableException;
+
+        RetryContext(
+                int retryIndex, int totalTimes, Class<? extends Throwable> repeatableException) {
+            this.retryIndex = retryIndex;
+            this.totalTimes = totalTimes;
+            this.repeatableException = repeatableException;
+        }
+
+        @Override
+        public String getDisplayName(int invocationIndex) {
+            return String.format(
+                    "RetryIndex: [%d], Test:[%d/%d]", invocationIndex, retryIndex, totalTimes);
```

Comment:
what is the difference between `invocationIndex` and `retryIndex?`
```suggestion
                    "Attempt [%d/%d]", retryIndex, totalTimes);
```

Reply:
They are supposed to be the same. But the invocationIndex is the actual index of this invocation from junit, the retryIndex is the index I plan to execute. I prefer to use the invocationIndex from junit.

Reply:
yes, I agree that we should use he invocationIndex

Reply:
After refactoring this part of code, I cannot get the invocationIndex in some class. So I have to change to use the planned `retryIndex`. ;)